manager/state/raft/raft.go

```diff
@@ -240,6 +243,51 @@ func NewNode(opts NodeOptions) *Node {
 	return n
 }
 
+// SetAddr provides the raft node's address. This can be used in cases where
+// opts.Addr was not provided to NewNode, for example when a port was not bound
+// until after the raft node was created.
+func (n *Node) SetAddr(ctx context.Context, addr string) error {
+	n.addrLock.Lock()
+	defer n.addrLock.Unlock()
+
+	n.opts.Addr = addr
+
+	if n.IsMember() {
+		if err := n.cluster.SetNodeAddr(n.Config.ID, addr); err != nil {
+			return err
+		}
+
+		// If the raft node is running, submit a configuration change
+		// with the new address.
+
+		// TODO(aaronl): Currently, this node must be the leader to
+		// submit this configuration change. This works for the initial
+		// use cases (single-node cluster late binding ports, or calling
+		// SetAddr before joining a cluster). In the future, we may want
+		// to support having a follower proactively change its remote
+		// address.
+
+		leadershipCh, cancel := n.SubscribeLeadership()
+		defer cancel()
+
+		isLeader := atomic.LoadUint32(&n.signalledLeadership) == 1
+		for !isLeader {
```

Comment:
As we discussed offline, currently only the leader can submit proposals such as configuration changes. For now, this `SetAddr` function is intended to address the narrow use case of a single-node cluster that didn't have ports bound before. Since the cluster only has one member, this node must be the leader, but there can be delays before that leadership is confirmed, for example at startup. This loop waits until the current node becomes the leader.

If/when we extend `SetAddr` to handle more general cases, we'll need some other approach here, such as an RPC we can use to tell the leader to update a node's address.